docs/src/tutorials/write-my-software-management-plugin.md

```diff
@@ -0,0 +1,393 @@
+Let's call the pre-defined argument such as `list`, `install`, and `remove` a **command** here. 
+As you can see from this example, a plugin should be an executable file 
+that accepts the commands and outputs to stdout and stderr in the defined JSON format. 
+Hence, you can implement a plugin in your preferred language.
+
+> **Important**: the Software Management Agent executes a plugin using `sudo` and as `tedge-agent` user.
+
+Here is the table of the commands that you can use in a plugin.
+
+|Command|Input arguments|Expected output|Description|
+|---|---|---|---|
+|list| - | JSON Lines |Returns the list of software modules that have been installed with this plugin.|
+|prepare| - | - |Executes the provided actions before a sequence of install and remove commands.|
+|finalize| - | - |Executes the provided actions after a sequence of install and remove commands.|
+|install| NAME [--version VERSION] [--file FILE] | - |Executes the action of installation.|
+|remove| NAME [--version VERSION] | - |Executes the action of uninstallation.|
+|update-list| COMMAND NAME [--version VERSION] [--file FILE] | - |Executes the list of `install` and `remove` commands.|
+
+The order of the commands invoked by the Software Management Agent is:
+`prepare` -> `update-list` or [`install`, `remove`] ->`finalize`
+
+> **info**: There is no guarantee of the order between `install` and `remove`.
+> If you need a specific order, use `update-list` command instead.
+
+In the following sections, we will dive into each command and other rules deeply.
+
+## Input, Output, and Errors
+
+Before we dive into each command, we should clarify the basic rules of plugins.
+
+### Input
+
+The command themselves and further required arguments must be given as command-line arguments.
+The only exception is `update-list`, which requires **stdin** input.
+
+### Output
+
+The **stdout** and **stderr** of the process running a plugin command are captured by the Software Management Agent.
+
+### Exit status
+
+The exit status of plugins are interpreted by sm-agent as follows:
+- **0**: success.
+- **1**: usage. The command arguments cannot be interpreted, and the command has not been launched.
+- **2**: failure. The command failed and there is no point to retry.
+- **3**: retry. The command failed but might be successful later (for instance, when the network will be back).
+
+## List
+
+The `list` command is responsible to return the list of the installed software modules.
+
+Rules:
+
+- This command takes no arguments.
+- The output must be in [the JSON Lines format](https://jsonlines.org/) including:
+  - **name**: the name of the software module, e.g. `mosquitto`.
+  This name is the name that has been used to install it and that needs to be used to remove it.
+  - **version**: the version currently installed.
+  This is a string that can only be interpreted in the context of the plugin.
+  
+Given that your plugin is named `myplugin`, then the Software Management Agent calls
```

Comment:
Might be better not to introduce another plugin name here. I would stick to the `docker` plugin.